feat : Clean Architecture + Layerd Architecture 적용

devNana222 opened the pull request:
특강 신청 프로그램의 기본 구조를 작성했습니다.

# application
- command sturcture (LectureCommand.java)를 활용하여 application <> presentation의 결합도를 낮췄습니다.
- command객체를 repository의 매개변수로 전달해 줍니다.

# domain
- Business계층에서 일어나는 예외처리를 LectureErrorCode로 정의했으며 이후 테스트 코드를 작성하며 추가할 예정입니다.
- JPA를 사용하기 때문에 Entity를 Domain에 해당 소스와 같이 설정하였고 강의의 정보가 담긴 Lecture / 강의 신청 목록을 불러오는 LectureHistory / 수강생의 정보가 담긴 Student 로 테이블을 구성하였습니다.

# infrastructure
- JpaRepository를 확장시킨 각 RepositoryInterface를 생성하였습니다.

# presentation
- 컨르롤러에 세가지 요구사항에 담긴 api를 작성하였습니다.
- 매개변수는 모두 command객체를 넘겨줌으로서 결합을 느슨하게 만들었습니다.



*리뷰필요한 항목* 
- Layerd Architecture + Clean Architecture 의 구조로 만들고 싶은데 잘못 된 부분이 있으면 의견 부탁드립니다. 
- 네이밍이 어색한 부분이 있을까요? 
- @Transactional 이라는 어노테이션이 있다고 배워서 사용해 봤는데 제대로 사용한 것이 맞을지, 해당 부분을 테스트 하려면 어떤식으로 테스트 코드를 짜야하는지 궁금합니다. 
- Command 패턴을 사용하고있는데 이부분을 테스트 하기 위해선 어떻게 유닛 테스트를 해야할지 모르겠습니다. 
-

Commit:
feat : Clean Architecture + Layerd Architecture 적용

특강 신청 프로그램의 기본 구조를 작성했습니다.

# application
- command sturcture (LectureCommand.java)를 활용하여 application <> presentation의 결합도를 낮췄습니다.
- command객체를 repository의 매개변수로 전달해 줍니다.

# domain
- Business계층에서 일어나는 예외처리를 LectureErrorCode로 정의했으며 이후 테스트 코드를 작성하며 추가할 예정입니다.
- JPA를 사용하기 때문에 Entity를 Domain에 해당 소스와 같이 설정하였고
강의의 정보가 담긴 Lecture / 강의 신청 목록을 불러오는 LectureHistory / 수강생의 정보가 담긴 Student 로 테이블을 구성하였습니다.

# infrastructure
- JpaRepository를 확장시킨 각 RepositoryInterface를 생성하였습니다.

# presentation
- 컨르롤러에 세가지 요구사항에 담긴 api를 작성하였습니다.
- 매개변수는 모두 command객체를 넘겨줌으로서 결합을 느슨하게 만들었습니다.

psam1017 commented:
## 리뷰 답변

```
Layerd Architecture + Clean Architecture 의 구조로 만들고 싶은데 잘못 된 부분이 있으면 의견 부탁드립니다.
네이밍이 어색한 부분이 있을까요?
```
네이밍에 어색하다고 생각한 건 없었구요. 아키텍처는 제가 감히 판단할 수준은 아닌 것 같지만..명백하게 이건 아니다! 싶은 건 없었어요.

```
@transactional 이라는 어노테이션이 있다고 배워서 사용해 봤는데 제대로 사용한 것이 맞을지, 해당 부분을 테스트 하려면 어떤식으로 테스트 코드를 짜야하는지 궁금합니다.
```
네 @Service 클래스에서 적절하게 사용되고 있는 것 같습니다. 참고로 @Transactional(readOnly = true) 라는 식으로 옵션을 줄 수도 있는데, 이렇게 하면 변경 감지를 통한 수정을 불가능하지만, 내부적인 동작으로 조회만 하는 경우의 쿼리를 수행할 때 약간 더 성능적인 이득을 얻을 수 있다고 해요.
그리고 지금 작성하신 코드에서는 괜찮지만, JPA 가 객체 그래프 탐색을 하기 위한 Fetch 를 할 때 트랜잭션 범위 밖에서는 탐색을 하지 못 하는 경우도 있어서, 그럴 경우만 유의하신다면 대부분의 경우에 사용하실 때는 무리가 없을 것 같습니다.

```
Command 패턴을 사용하고있는데 이부분을 테스트 하기 위해선 어떻게 유닛 테스트를 해야할지 모르겠습니다.
```
[LectureCommand.Apply](https://github.com/devNana222/hhplus0010/pull/1/files#diff-813e4a97fa2908d3c971ab08119b1409cd47e418b8bb0f472b80b7d7c427741dR9)
저라면 LectureCommand.Apply 에 studentId, lectureId 에 null 을 전달하고 없을 때 예외가 발생한다는 테스트만 해볼 것 같아요! 저도 그 이외에 큰 테스트 필요를 느끼지는 못할 것 같습니다.
만약 공개된 메서드들인 isAvailable 이나 from 등에서도 잘 작동하는지 테스트한다면 할 수 있을 것 같아요. 처음 만들 때 stub 으로 return false, return null 이렇게 두고 나중에 만들어야지~ 하고 까먹었다가 나중에 통합테스트에서 걸릴 수도 있는 게 걱정이 된다면요.

## 추가

[@Data 사용](https://github.com/devNana222/hhplus0010/pull/1/files#diff-b1653915cd856bf56554c650c65a48aeb38eaea008f61562b1b8c56eb747230aR12)

제가 Spring 배우면서 배운 몇 개 중에 `@Data` 를 신중히 사용하라는 내용이 있었습니다.
엔티티 중에서 변경되면 안 될 필드에 대한 Setter 가 만들어져서 비즈니스 규칙에 대한 오해, 또는 setter 남용을 일으킬 수 있어서 `@Getter` 와 함께 필요한 lombok 을 직접 추가해주면 좋을 것 같습니다.

Commit:
feat : 피드백 적용, Default 기능 구현

# 요구사항

- 아키텍처 준수를 위한 애플리케이션 패키지 설계
- 특강 도메인 테이블 설계 및 목록/신청 등 기본 기능 구현
- 각 기능에 대한 **단위 테스트** 작성

> 사용자 회원가입/로그인 기능은 구현하지 않습니다.

## 단위테스트 항목
- 각 application > Service에 대한 단위 테스트 진행하였습니다.
- controller api별 단위테스트 진행하였습니다.

## 불필요한 항목 제거 & 테스트 하며 나온 오류 내용 수정
- 사용되지않는 import 제거하였습니다.
- 유효성 검사 적용하였습니다. (command 안에서 String으로 들어온 데이터 LocalDate로 변환)
- studentService라는 명칭이 헷갈릴 것 같아 "lectureHistory"로 변경하였습니다.

devNana222 commented on src/main/java/com/tdd/application/LectureApplyService.java:
```diff
@@ -37,19 +36,31 @@ public LectureApplyService(LectureRepository lectureRepository
-    /**
-     *
-    * */
+    @Transactional(readOnly = true)
```
지난번 피드백으로 readOnly=true를 적용하였습니다. 
실제 적용이 되는지는 통합테스트 할 때 확인해봐야겠습니다 ㅎㅎ;

devNana222 commented on src/main/java/com/tdd/application/command/LectureCommand.java:
```diff
@@ -20,19 +22,21 @@ public static Apply from(Long studentId, Long LectureId) {
-    public static class Available {
-        private boolean available;
+    public static class Date {
+        @Pattern(regexp = "^$|\\\\d{4}-\\\\d{2}-\\\\d{2}", message = "날짜 형식이 잘못되었습니다. yyyy-MM-dd 형식이어야 합니다.")
```
pattern을 command에 작성하는게 맞는건가 싶긴한데.. 더 좋은 위치가 있으면 알려주세요! 
그리고 여기서 message 대신 businessException을 적용하고싶으면 어떻게 변경하면좋을까요?

devNana222 commented on src/main/java/com/tdd/infrastructure/jpaEntity/LectureJpaRepository.java:
```diff
@@ -0,0 +1,27 @@
+    @Query("SELECT l FROM Lecture l WHERE l.enrollStartDate = :date AND l.capacity > 0")
+    List<Lecture> findAvailableLecturesByDate(LocalDate date);
+
+    @Lock(LockModeType.PESSIMISTIC_WRITE)
```
비관적 락을 이렇게 거는 것이 맞을까요? lectureRepository에 거는것도 맞을지 궁금합니다.

Commit:
feat : Default기능 구현 2

1. 불필요한 주석 제거
2. 일부 코드 리팩토링
3. capacity 체크 방식 변경
* AS-IS : history에서 lecture_id로 개수를 조회해서 capacity랑 비교하는 방식
* TO-BE : capacity를 초기 설정해 두고 특강신청 성공 시 1명식 감소하는 형태

chawchaw commented on src/main/java/com/tdd/infrastructure/jpaEntity/LectureHistoryJpaRepository.java:
```diff
@@ -0,0 +1,20 @@
+package com.tdd.infrastructure.jpaEntity;
```
jpaEntity 라는 패키지이름이 어색한 것  같아요.
실제로 여기에는 JpaRepository 들의 기능을 사용하는데
Entity 가 있을 거라고 예상하고 들어왔는데 다른 내용들이 있어서 이해하는데 시간이 좀 걸렸어요

livelysb replied:
그리고, java에서 패키지에 는 대문자를 사용하지 않아요~
